### PR TITLE
V3 permutation and task manager 

### PR DESCRIPTION
**Pull request by sallyi95**
Hi Nakib,

This PR contains a task_manager module to coordinate task distribution. Associated test for the task manager are added under test/test_task_manager.f90.
It also includes V3 using permutation. 

Looking to your feedback/suggestions.
Best,
Sally

### CODE REVIEW

File: src/task_manager.f90 [hunk 1]
```diff
@@ -0,0 +1,94 @@
+module task_manager_module
+        !! This module help to manage and distribute a set of tasks across a specified number of batches
```
Comment (nakib) [inline src/task_manager.f90]: The indentation of the comment here is too much.

File: src/task_manager.f90 [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+     private
+
+     integer :: num_batches = 1
+     integer, allocatable :: batch_info(:, :)
+  
+     contains
+       procedure, public :: distribute_load
+       procedure, public :: print_report
+       procedure, public :: get_num_batches
+       procedure, public :: get_batch_range
```
Comment (nakib) [inline src/task_manager.f90]: Suggestion: These 4 lines could be written as `procedure, public :: distribute_load, print_report, ...`

File: src/task_manager.f90 [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  end type task_manager
+
+contains
+
+  subroutine distribute_load(self, num_tasks, num_batches)
+    class(task_manager), intent(out) :: self
```
Comment (nakib) [inline src/task_manager.f90]: Suggestion: Add explanation about what this subroutine does. The docstring should start with `!!`.

File: src/task_manager.f90 [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    !how many tasks are left over
+    residual_task = mod(num_tasks, num_batches)
+
+    !the first batch index start is 1 always
+    start_idx = 1
+
+    !record the start and the end index and the size of each batch
+    do ibatch = 1, self%num_batches
+
```
Comment (nakib) [inline src/task_manager.f90]: Remove this empty line.

File: src/task_manager.f90 [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+       !save the start, end, and size of the batch
+       self%batch_info(ibatch, 1) = start_idx
+       self%batch_info(ibatch, 2) = end_idx
+       self%batch_info(ibatch, 3) = end_idx - start_idx + 1
+
+       !here the starting point for the next batch
+       start_idx = end_idx + 1
+    end do
+
```
Comment (nakib) [inline src/task_manager.f90]: Remove this empty line.

File: src/task_manager.f90 [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    integer, intent(in) :: num_tasks, num_batches
+
+    integer :: batch_size, residual_task, ibatch, start_idx, end_idx
+
+    !this handles cases where batches are more than tasks
+    self%num_batches = min(num_tasks, num_batches)
+    allocate(self%batch_info(self%num_batches, 3))
+
+    !divide the total number of tasks in num_batches (subtasks)
+    batch_size = num_tasks / num_batches
```
Comment (nakib) [inline src/task_manager.f90]: The styleguide suggests writing `a/b` instead of `a / b`.

File: src/task_manager.f90 [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+       self%batch_info(ibatch, 3) = end_idx - start_idx + 1
+
+       !here the starting point for the next batch
+       start_idx = end_idx + 1
+    end do
+
+  end subroutine distribute_load
+
+  subroutine print_report(self)
+    class(task_manager), intent(in) :: self
```
Comment (nakib) [inline src/task_manager.f90:77]: Suggestion: Add documentation.

File: src/task_manager.f90 [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+    if (this_image() == 1) then
+            !if (.true.) then
```
Comment (nakib) [inline src/task_manager.f90]: Remove this comment.

File: src/task_manager.f90 [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    residual_task = mod(num_tasks, num_batches)
+
+    !the first batch index start is 1 always
+    start_idx = 1
+
+    !record the start and the end index and the size of each batch
+    do ibatch = 1, self%num_batches
+
+    !first check if this batch should receive one of the extra tasks
+       if (ibatch <= residual_task) then
```
Comment (nakib) [inline src/task_manager.f90]: The styleguide suggests writing `if(clause)` instead of `if (clause)`.

File: src/task_manager.f90 [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+       !here the starting point for the next batch
+       start_idx = end_idx + 1
+    end do
+
+  end subroutine distribute_load
+
+  subroutine print_report(self)
+    class(task_manager), intent(in) :: self
+
+    if (this_image() == 1) then
```
Comment (nakib) [inline src/task_manager.f90]: The styleguide suggests writing `if(clause)` instead of `if (clause)`.

File: src/task_manager.f90 [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+       print*, "Number of batches:", self%num_batches
+       print*, "First index of batch:", self%batch_info(:, 1)
+       print*, "Last index of batch:", self%batch_info(:, 2)
+       print*, "Size of batch:", self%batch_info(:, 3)
+       print*, '========'
+    end if
+  end subroutine print_report
+
+  pure integer function get_num_batches(self)
+          !!returns the number of batches
```
Comment (nakib) [inline src/task_manager.f90]: The indentation of the docstring is wrong.

File: src/task_manager.f90 [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+    class(task_manager), intent(in) :: self
+    get_num_batches = self%num_batches
+  end function get_num_batches
+
+  pure function get_batch_range(self, batch) result(batch_range)
+          !!gives the range and size of a batch as a 3-element array: (start, end, size).
```
Comment (nakib) [inline src/task_manager.f90]: The indentation of the docstring is wrong.

File: src/task_manager.f90 [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+       print*, "Size of batch:", self%batch_info(:, 3)
+       print*, '========'
+    end if
+  end subroutine print_report
+
+  pure integer function get_num_batches(self)
+          !!returns the number of batches
+
+    class(task_manager), intent(in) :: self
+    get_num_batches = self%num_batches
```
Comment (nakib) [inline src/task_manager.f90:93]: Add an empty line before line 81.

File: src/task_manager.f90 [hunk 14]
```diff
@@ -0,0 +1,94 @@
+module task_manager_module
+        !! This module help to manage and distribute a set of tasks across a specified number of batches
+
+  implicit none
+  private
```
Comment (nakib) [inline src/task_manager.f90:8]: Add an empty line after the `implicit none` statement.

File: src/task_manager.f90
Comment (nakib) [inline src/task_manager.f90:1]: It would be best to change all integers to `integer(i64)`.

**Review by nakib — CHANGES_REQUESTED** [review]

**Comment by sallyi95** [comment]
Many thanks for your helpful suggestions.
I’ve implemented the recommended changes and happy to tweak further if needed.

File: src/task_manager.f90 [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+       print*, "Number of batches:", self%num_batches
+       print*, "First index of batch:", self%batch_info(:, 1)
+       print*, "Last index of batch:", self%batch_info(:, 2)
+       print*, "Size of batch:", self%batch_info(:, 3)
+       print*, '========'
+    end if
+  end subroutine print_report
+
+  pure integer function get_num_batches(self)
+    !!returns the number of batches
```
Comment (nakib) [inline src/task_manager.f90]: Suggestion: Capitalize first letter, use a space after '!!', and use period to mark the end of sentence.

File: src/task_manager.f90 [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+    class(task_manager), intent(in) :: self
+
+    get_num_batches = self%num_batches
+  end function get_num_batches
+
+  pure function get_batch_range(self, batch) result(batch_range)
+    !!gives the range and size of a batch as a 3-element array: (start, end, size).
```
Comment (nakib) [inline src/task_manager.f90]: Suggestion: Capitalize first letter and use a space after '!!'.

File: src/task_manager.f90 [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    residual_task = mod(num_tasks, self%num_batches)
+
+    !the first batch index start is 1 always
+    start_idx = 1
+    !record the start and the end index and the size of each batch
+    do ibatch = 1, self%num_batches
+       !first check if this batch should receive one of the extra tasks
+       if(ibatch <= residual_task) then
+          end_idx = start_idx + batch_size
+          !for the batches after the extra ones, assign exactly the batch_size tasks
```
Comment (nakib) [inline src/task_manager.f90]: Is this comment for the `else` branch? If yes, can you write it next to the `else` statement?

File: src/task_manager.f90 [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+    !divide the total number of tasks in num_batches (subtasks)
+    batch_size = num_tasks/self%num_batches
+
+    !how many tasks are left over
+    residual_task = mod(num_tasks, self%num_batches)
+
+    !the first batch index start is 1 always
+    start_idx = 1
+    !record the start and the end index and the size of each batch
```
Comment (nakib) [inline src/task_manager.f90]: Add an empty line after line 47.

File: src/task_manager.f90 [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+contains
+
+  subroutine distribute_load(self, num_tasks, num_batches)
+    !!Partitions a total number of tasks into number of batches
+    !!
+    !!balancing the work as evenly as possible.
+    !!the first few batches can receive one additional task if the 
+    !!total number of task is not divisible evenly by the number of batches
+    !!the number of batches used is limited to min(num_tasks, num_batches), mean no batch is left empty
```
Comment (nakib) [inline src/task_manager.f90]: Suggestion: Use a space after `!!`. Write the docstring in standard English if possible, completing sentences properly and using punctuation, as needed.

File: src/task_manager.f90 [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+  private
+  public :: task_manager
+
+  type :: task_manager
+     private
+
+     integer(i64) :: num_batches = 1
+     integer(i64), allocatable :: batch_info(:, :)
+
+   contains
```
Comment (nakib) [inline src/task_manager.f90:21]: Suggestion: add an empty line after the `contains` statement.

File: src/task_manager.f90 [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+  !! This module help to manage and distribute a set of tasks across a specified number of batches
+
+  use precision, only: i64
+
+  implicit none
+
+  private
+  public :: task_manager
+
+  type :: task_manager
```
Comment (nakib) [inline src/task_manager.f90:11]: Needs to be documented.

File: src/task_manager.f90 [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+     private
+
+     integer(i64) :: num_batches = 1
+     integer(i64), allocatable :: batch_info(:, :)
```
Comment (nakib) [inline src/task_manager.f90]: Each member should be documented. See, for example, https://github.com/nakib/elphbolt/blob/develop/src/particle.f90

File: test/test_task_manager.f90
Comment (nakib) [inline test/test_task_manager.f90:1]: Since you are using `testify`, might as well write some unit tests, making sure to test for edge cases.

**Review by nakib — CHANGES_REQUESTED** [review]

File: src/task_manager.f90 [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+   contains
+
+     procedure, public :: distribute_load, print_report, get_num_batches, get_batch_range
+  end type task_manager
+
+contains
+
+  subroutine distribute_load(self, num_tasks, num_batches)
+    !! Partitions a total number of tasks into number of batches
+    !!
```
Comment (nakib) [inline src/task_manager.f90]: The line break at line 30 is unnecessary.

File: src/task_manager.f90 [hunk 24]
```diff
@@ -0,0 +1,107 @@
+module task_manager_module
+  !! This module help to manage and distribute a set of tasks across a specified number of batches.
```
Comment (nakib) [inline src/task_manager.f90]: Change to "!! Module containing the data type related to task batching."

File: src/task_manager.f90 [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+  use precision, only: i64
+
+  implicit none
+
+  private
+  public :: task_manager
+
+  type :: task_manager
+     !! Container for task batching strategy that assign tasks to batches.
+     private
```
Comment (nakib) [inline src/task_manager.f90:14]: Need a line break after line 12.

File: src/task_manager.f90 [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+
+     integer(i64) :: num_batches 
+     !! This is the number of batches.
```
Comment (nakib) [inline src/task_manager.f90]: Just write "!! Number of batches."

File: src/task_manager.f90 [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+  public :: task_manager
+
+  type :: task_manager
+     !! Container for task batching strategy that assign tasks to batches.
+     private
+
+     integer(i64) :: num_batches 
+     !! This is the number of batches. 
+     integer(i64), allocatable :: batch_info(:, :)
+     !! This contains the task batching informations.
```
Comment (nakib) [inline src/task_manager.f90]: Just write "!! Task batching information."

File: src/task_manager.f90 [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+
+  type :: task_manager
+     !! Container for task batching strategy that assign tasks to batches.
+     private
+
+     integer(i64) :: num_batches 
+     !! This is the number of batches. 
+     integer(i64), allocatable :: batch_info(:, :)
+     !! This contains the task batching informations. 
+     !! First axis is runs over number of batches. Second axis gives the first index, last index, and size of each batch, respectively.
```
Comment (nakib) [inline src/task_manager.f90]: The first sentence is grammatically incorrect.

File: src/task_manager.f90 [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+
+contains
+
+  subroutine distribute_load(self, num_tasks, num_batches)
+    !! Partitions a total number of tasks into number of batches
+    !!
+    !! balancing the work as evenly as possible.
+    !! The first few batches can receive one additional task if the 
+    !! total number of task is not divisible evenly by the number of batches.
+    !! The number of batches used is limited to min(num_tasks, num_batches), which means no batch is left empty.
```
Comment (nakib) [inline src/task_manager.f90]: Can simply say something like "!! We use a shuffling algorithm here." The last line is fine.

**Review by nakib — CHANGES_REQUESTED** [review]

**Review by nakib — APPROVED** [review]
